google/flutter-mediapipe
# [WIP] FFI Implementation

Adds a MediaPipe implementation that talks directly to C++ via FFI. (Technically, indirectly and through a C intermediary, but still fairly directly in comparison to method channels.)

C bridge code uses Dart's native FFI functionality and `ffigen` to automatically generate the required bindings from C headers pulled from the MediaPipe repository.

Includes:
* Beginnings of a shared `mediapipe_core` library with Dart classes for low level classification objects common to many MediaPipe Tasks
  * Unit tests for the above
* Beginnings of `mediapipe_text` library with a `TextClassifier` class that knows how to ask MediaPipe to classify text
  * Integration tests for the above (the `TextClassifier` uses real FFI to call into a faked implementation in C)

TODOs (in or outside of this PR):
* Automate updates to header files from MediaPipe (these are rumored to be exceedingly rare)
* Vendor MediaPipe libraries (google/flutter-mediapipe#9)
* Explore web implementation (which will use JS interop instead of FFI)

**TESTING**

To text this PR locally, complete the following steps (on macOS):

1. In a checkout of `google/mediapipe` with the latest internal changes pulled down, run this build command:

```sh
bazel build --config darwin_arm64 --strip always -c opt --define MEDIAPIPE_DISABLE_GPU=1 mediapipe/tasks/c/text/text_classifier:libtext_classifier.dylib
```

2. Copy the resulting `libtext_classifier.dylib` file (found in the `bazel-out` directory) to this repository at the `packages/mediapipe-task-text/examples/assets` directory. (Hint: It should be right next to `bert_classifier.tflite`, which is in source control.)

3. Run the example app and click the floating action button to classify text.
